A first working prototype of saving model results in mongo

knoxsp opened the pull request.

Commit:
A first working prototype of saving model results in mongo

pmslavin reviewed (changes requested):
A couple of minor observations:
* A few places where try/except catches too-general an exception or all exceptions
* Some functions processing recorder where the complexity requires comments

Also, some kind of round-trip test for the MongoResultsProcessor would demonstrate correctness(?)

Commit:
Store results in s3 at the end of processing the mongo results

Commit:
Add a flag (not fully configureable yet) which stores the results directly to mongo instead of storing them in h5

Commit:
Merge branch '58-store-results-in-a-mongo-db-if-available' of github.com:pywr/hydra-pywr into 58-store-results-in-a-mongo-db-if-available

Commit:
Make arguments to run_network_scenario consistent

Commit:
Merge branch 'master' into 58-store-results-in-a-mongo-db-if-available

Commit:
Remove duplicated code from master merge

Commit:
Output the scenrio ID when upload complete

Commit:
Use pyproject.toml instead of setup.py

Commit:
Include sub-modules in pyproject toml

Commit:
Two updates:
1: Add a flag (bases on an environmenta variable) to store data in a
node-centric way in h5 instead of attribute-centric.
2: Add a flag to turn off adding of the simulated flow and volume
parameters, which overrides the 'record_nodes' feature (which is still
present).

pmslavin commented on hydra_pywr/cli.py:
```diff
@@ -8,6 +8,7 @@
 from . import runner
 from . import exporter
 from . import importer
+
```
```suggestion
```

pmslavin commented on hydra_pywr/resultsprocessor/hydra.py:
```diff
@@ -0,0 +1,402 @@
+            for ra_id, metadata in returned_new_ids.items():
+                new_ids[tuple(metadata)] = ra_id
+
+            for i, new_ra in enumerate(chunk):
```
`i` is unused here, can just `for new_ra in chunk:`

pmslavin commented on hydra_pywr/resultsprocessor/hydra.py:
```diff
@@ -0,0 +1,402 @@
+            in the template
+        """
+        attr_name_map = {}
+        for templatetype in self.hydra_template.templatetypes:
```
`make_attr_unit_map` checks for an existing template but `make_attr_name_map` doesn't.

pmslavin commented on pyproject.toml:
```diff
@@ -0,0 +1,55 @@
+Homepage = "https://github.com/pywr/hydra-pywr"
+Documentation = "https://hydraplatform.org"
+Repository = "https://github.com/pywr/hydra-pywr"
+"Bug Tracker" = "https://github.com/pywr/hydra-pywr/issues"
```
There's an `issues` (and some aliases) [well-known label](https://packaging.python.org/en/latest/specifications/well-known-project-urls/#well-known-labels) instead of "Bug Tracker".

pmslavin reviewed (approved):
Several comments, suggestions, and requests for clarifications, but nothing blocking merge.

Commit:
add boto3 to toml

Commit:
Merge branch 'master' into 58-store-results-in-a-mongo-db-if-available

Commit:
move parser import to the code to avoid clashes with circular importing

Commit:
1: Add a --cache flag to use an already downloaded model if available
2: Add a --dry-run flag to not save data back to hydra after the model
   is run, to test model runs but not save the results
3: Add a --update flag which when set doesn't delete resource scenarios
   before saving results
4: Add support for saving simulated volume on virtual storage nodes and
   simulated flow on aggregated nodes
